x/tokenomics/keeper/token_logic_modules.go

```diff
@@ -0,0 +1,443 @@
+	if err := k.bankKeeper.MintCoins(ctx, tokenomictypes.ModuleName, newMintCoins); err != nil {
+		return tokenomicstypes.ErrTokenomicsModuleMintFailed.Wrapf(
+			"minting %s to the tokenomics module account: %v", newMintCoins, err)
+	}
+	logger.Info(fmt.Sprintf("minted (%v) coins in the tokenomics module", newMintCoins))
+
+	// Send a portion of the rewards to the application
+	if err := k.sendRewardsToAccount(ctx, application.Address, settlementAmtFloat, MintAllocationApplication); err != nil {
+		return err
+	}
+
+	// Send a portion of the rewards to the supplier
+	if err := k.sendRewardsToAccount(ctx, supplier.Address, settlementAmtFloat, MintAllocationSupplier); err != nil {
+		return err
+	}
+
+	// Send a portion of the rewards to the DAO
+	if err := k.sendRewardsToAccount(ctx, k.GetAuthority(), settlementAmtFloat, MintAllocationDAO); err != nil {
+		return err
+	}
+
+	// Send a portion of the rewards to the source owner
+	if err := k.sendRewardsToAccount(ctx, service.OwnerAddress, settlementAmtFloat, MintAllocationSourceOwner); err != nil {
+		return err
+	}
+
+	// Send a portion of the rewards to the block proposer
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	proposerAddr := cosmostypes.AccAddress(sdkCtx.BlockHeader().ProposerAddress).String()
+	if err := k.sendRewardsToAccount(ctx, proposerAddr, settlementAmtFloat, MintAllocationProposer); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// sendRewardsToAccount sends (settlementAmtFloat * allocation) tokens from the
+// tokenomics module account to the specified address.
+func (k Keeper) sendRewardsToAccount(
+	ctx context.Context,
+	addr string,
+	settlementAmtFloat *big.Float,
+	allocation float64,
+) error {
+	logger := k.Logger().With("method", "mintRewardsToAccount")
+
+	accountAddr, err := cosmostypes.AccAddressFromBech32(addr)
+	if err != nil {
+		return err
+	}
+
+	coinsToAccAmt, _ := new(big.Float).Mul(settlementAmtFloat, big.NewFloat(allocation)).Int64()
+	coinsToAcc := sdk.NewCoins(cosmostypes.NewCoin(volatile.DenomuPOKT, math.NewInt(coinsToAccAmt)))
+	if err := k.bankKeeper.SendCoinsFromModuleToAccount(
+		ctx, suppliertypes.ModuleName, accountAddr, coinsToAcc,
+	); err != nil {
+		return err
+	}
+	logger.Info(fmt.Sprintf("sent (%v) coins from the tokenomics module to the account with address %q", coinsToAcc, addr))
+
+	return nil
+}
+
+func (k Keeper) handleOverservicedApplication(
+	ctx context.Context,
+	application *apptypes.Application,
+	settlementCoins cosmostypes.Coin,
+) (
+	newSettlementCoins cosmostypes.Coin,
+	err error,
+) {
+	logger := k.Logger().With("method", "handleOverservicedApplication")
+	// over-serviced application
+	logger.Warn(fmt.Sprintf(
+		"THIS SHOULD NEVER HAPPEN. Application with address %s needs to be charged more than it has staked: %v > %v",
+		application.Address,
+		settlementCoins,
+		application.Stake,
+	))
+
+	// TODO_MAINNET(@Olshansk, @RawthiL): The application was over-serviced in the last session so it basically
```

Comment:
PTAL at https://www.notion.so/buildwithgrove/Off-chain-Application-Stake-Tracking-6a8bebb107db4f7f9dc62cbe7ba555f7 it's maybe worth mentioning in this comment.

Reply:
Perfect!